### PR TITLE
Modal

### PR DESCRIPTION
**Pull request by xutony05**
Image says it all

image
image

Modal for when user make 3 mistakes. User can restart, quit, change to prev unlocked level, or switch algo.

### CODE REVIEW

File: client/src/components/LevelLayout.tsx [hunk 1]
```diff
@@ -46,6 +57,19 @@ const LevelLayout: React.FC<Props> = ({
     didWin();
   }
 
+  const [algo, setAlgo] = useState<number>(0);
+  const [level, setLevel] = useState<number>(0);
+  const [customRestart, setCustomRestart] = useState<boolean>(false);
+  const navigate = useNavigate();
+
+  function generateLevels() {
```
Comment (garethlau) [inline client/src/components/LevelLayout.tsx:65]: For performance, it's not ideal to have function calls return DOM elements as the function will be called on each render even in the return of the function stays the same. Generally speaking, this is a bad practice, see [here](https://kentcdodds.com/blog/dont-call-a-react-function-component).

Instead, consider using `useMemo`. 

This is super, duper not necessary though. I wouldn't be peeved if you merged as is.

File: client/src/components/LevelLayout.tsx [hunk 2]
```diff
@@ -75,6 +99,113 @@ const LevelLayout: React.FC<Props> = ({
...
+                      {name}
+                    </option>
+                  ))}
+                </Select>
+                <Select
+                  onChange={(e) => setLevel(parseInt(e.target.value))}
+                  defaultValue={"0"}
+                  marginTop={2}
+                >
+                  {generateLevels()}
```
Comment (garethlau) [inline client/src/components/LevelLayout.tsx:135]: See related comment about the `generateLevels` method.

File: client/src/components/LevelLayout.tsx [hunk 3]
```diff
@@ -75,6 +99,113 @@ const LevelLayout: React.FC<Props> = ({
...
+                <Button
+                  colorScheme="red"
+                  mr={3}
+                  onClick={() => {
+                    game.restartLevel();
+                    navigate("/");
+                  }}
+                >
+                  Quit
+                </Button>
```
Comment (garethlau) [inline client/src/components/LevelLayout.tsx:171]: It seems there is a lot of duplicated code. Consider using the `customRestart` boolean at a lower granularity to reduce duplicate code.

**Review by garethlau — APPROVED** [review]
Looks really good @xutony05 , great work. I'm approving this but take a look at some of the comments. I have no problem if you don't address them, they aren't a big deal.

File: client/src/components/LevelLayout.tsx [hunk 4]
```diff
@@ -1,17 +1,28 @@
...
+  ModalBody,
   useToast,
 } from "@chakra-ui/react";
 import { useGame } from "@/context/GameContext";
 import { TOOLBAR_HEIGHT } from "@/constants";
 import BoxesContainer from "./BoxesContainer";
 import StepValidation from "./StepValidation";
 import Instructions from "./Instructions";
+import { OPTIONS } from "@/constants";
+import { useNavigate } from "react-router";
```
[the remark this answers is not in the document]
Comment (xutony05) [reply client/src/components/LevelLayout.tsx:25]: fixed

File: client/src/components/LevelLayout.tsx [hunk 5]
```diff
@@ -46,6 +57,19 @@ const LevelLayout: React.FC<Props> = ({
...
   }
 
+  const [algo, setAlgo] = useState<number>(0);
+  const [level, setLevel] = useState<number>(0);
+  const [customRestart, setCustomRestart] = useState<boolean>(false);
+  const navigate = useNavigate();
+
+  function generateLevels() {
+    let items = [];
+    for (var i = 0; i <= game.maxLevelSeen; i++) {
```
[the remark this answers is not in the document]
Comment (xutony05) [reply client/src/components/LevelLayout.tsx:67]: fixed

File: client/src/components/LevelLayout.tsx (hunk 1 above)
Comment (xutony05) [reply client/src/components/LevelLayout.tsx:65]: updated

File: client/src/components/LevelLayout.tsx (hunk 3 above)
Comment (xutony05) [reply client/src/components/LevelLayout.tsx:171]: Very good point. Will keep the same for now, less coupling ;) easier to modify

**Pull request merged**